lib/datadog/tracing/distributed/datadog.rb

```diff
@@ -0,0 +1,150 @@
+        # DEV: The same comment applies to the {.extract_tags}.
+        def inject_tags(digest, data)
+          return if digest.trace_distributed_tags.nil? || digest.trace_distributed_tags.empty?
+
+          if ::Datadog.configuration.tracing.x_datadog_tags_max_length <= 0
+            active_trace = Tracing.active_trace
+            active_trace.set_tag('_dd.propagation_error', 'disabled') if active_trace
+            return
+          end
+
+          encoded_tags = DatadogTagsCodec.encode(digest.trace_distributed_tags)
+
+          if encoded_tags.size > ::Datadog.configuration.tracing.x_datadog_tags_max_length
+            active_trace = Tracing.active_trace
+            active_trace.set_tag('_dd.propagation_error', 'inject_max_size') if active_trace
+
+            ::Datadog.logger.warn(
+              "Failed to inject x-datadog-tags: tags are too large (size:#{encoded_tags.size} " \
+                "limit:#{::Datadog.configuration.tracing.x_datadog_tags_max_length}). This limit can be configured " \
+                'through the DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH environment variable.'
+            )
+            return
+          end
+
+          data[@tags] = encoded_tags
+        rescue => e
+          active_trace = Tracing.active_trace
+          active_trace.set_tag('_dd.propagation_error', 'encoding_error') if active_trace
+          ::Datadog.logger.warn(
+            "Failed to inject x-datadog-tags: #{e.class.name} #{e.message} at #{Array(e.backtrace).first}"
+          )
+        end
+
+        # Import `x-datadog-tags` tags as trace distributed tags.
+        # Only tags that have the `_dd.p.` prefix are processed.
+        def extract_tags(data)
+          tags = data[@tags]
+          return if !tags || tags.empty?
+
+          if ::Datadog.configuration.tracing.x_datadog_tags_max_length <= 0
+            active_trace = Tracing.active_trace
+            active_trace.set_tag('_dd.propagation_error', 'disabled') if active_trace
+            return
+          end
+
+          if tags.size > ::Datadog.configuration.tracing.x_datadog_tags_max_length
+            active_trace = Tracing.active_trace
+            active_trace.set_tag('_dd.propagation_error', 'extract_max_size') if active_trace
+
+            ::Datadog.logger.warn(
+              "Failed to extract x-datadog-tags: tags are too large (size:#{tags.size} " \
+                "limit:#{::Datadog.configuration.tracing.x_datadog_tags_max_length}). This limit can be configured " \
+                'through the DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH environment variable.'
+            )
+            return
+          end
+
+          tags_hash = DatadogTagsCodec.decode(tags)
+          # Only extract keys with the expected Datadog prefix
+          tags_hash.select! do |key, _|
+            key.start_with?(Tracing::Metadata::Ext::Distributed::TAGS_PREFIX) && key != EXCLUDED_TAG
+          end
```

Comment:
Minor: Consider moving this inside the `DatadogTagsCodec`? It's kinda weird that the `DatadogTagsCodec` returns things that are not Datadog tags and makes the caller do the work of actually filtering out non-datadog stuff.

Reply:
I think this is a hairy point: `Distributed::TAGS_PREFIX` only referrers to `_dd.p.*` tags, which is a subset of all `_dd.*` tags. The `DatadogTagsCodec` is meant to parse any tags in our custom Datadog-created format, but doesn't care if they are `_dd.p.*` or not.

Reply:
Aaaaahhhh I see. Thanks for clarifying, I had indeed missed that.